project-selector: hide exclusive upstreams/downstreams

scarlettperry opened the pull request:
### Description
PR adds support for hiding exclusive upstreams/downstreams of a user-owned/custom project when the latter are unchecked.

PR flow derives the hidden status based on the checked value (ie `true`/`false`) of the user-owned/custom project. In this way, we don't need to handle the flow in the `actions` as a the hidden status of an upstream/downstream is influenced by various user actions (the toggle, the only button, the checkbox, if a new project is added, etc).

### Testing Performed
locally with real API data

### TODOs
- [x] hide an upstream/downstream if all unchecked projects share it

### Questions:
~1. how do we want to handle the switch status b/c we preserve the checked status of hidden projects. we can simply filter by "if checked but not hidden". the `deriveSwitchStatus` helper is used in other parts tho so we'll need to create a helper for the non-switch status use cases.~ **in latest commit, we pass the derived data to the components so the project count already correctly reflects what's _not_ hidden**

~2. how do we want to alert the user to the number of hidden projects. we currently display total number of projects in the chip; we can simply subtract the number of hidden from that project but should we let the users know that some projects have been hidden?~ **we can revisit after collecting user feedback on the hiding behavior**

scarlettperry commented on frontend/workflows/projectSelector/src/project-group.tsx:
```diff
@@ -153,51 +156,54 @@ const ProjectGroup: React.FC<ProjectGroupProps> = ({ title, group, displayToggle
+          {Object.keys(state[group]).map(
+            key =>
+              !deriveHiddenStatus(state, group, key) && (
+                <StyledMenuItem key={key}>
```
no changes from this point on; the rest are just indentation updates b/c of the addition of line `161`

scarlettperry commented:
@danielhochman noted the main todo in the PR description and other less urgent ones in the PR but the e2e flow of hiding upstreams/downstreams there

scarlettperry commented on frontend/workflows/projectSelector/src/helpers.tsx:
```diff
@@ -96,7 +103,75 @@ const exclusiveProjectDependencies = (
+
+  // no unchecked projects
+  if (unchecked.length === 0) {
+    updateHidden(state, group, key, false);
```
this is buggy/doesn't always accurately reflect current state of hidden projects. still thinking about how to properly update the global state

scarlettperry replied:
actually the bug was with how I was handling some edge cases w.r.t hiding an upstream/downstream if all its parent projects are unchecked. will push up the fixes

danielhochman commented on frontend/workflows/projectSelector/src/helpers.tsx:
```diff
@@ -47,6 +48,12 @@ const updateGroupstate = (
 
+const updateHidden = (state: State, group: Group, project: string, hidden: boolean): State => {
+  const newState = { ...state };
+  newState[group][project].hidden = hidden;
```
if it's derived you'll remove it from the state entirely and only use it when rendering. should not be in both places.

scarlettperry replied:
yeah I did it like this so that we'd have a way to know whether upstream/downstream project data should be passed to cards (ie "if checked but not hidden"). 🤔  perhaps I can just do one - use the `deriveHiddenStatus` to only update the hidden state and then check the `hidden` field to decide whether to render

danielhochman replied:
the effect that will update the parent context is in the component, so it will have access to which projects are hidden or not when it updates the parent context

scarlettperry replied:
updated PR to have the derived computation in the project-selector component. i also tested this on your `dash-data-flow` branch and it worked nicely. in the new useEffect hook to update selected projects, I used the `derivedState` value.

scarlettperry commented on frontend/workflows/projectSelector/src/project-selector.tsx:
```diff
@@ -114,6 +114,10 @@ const ProjectSelector = ({ children }) => {
 
+  // computes the final state for rendering across other components
+  // (ie. filters out upstream/downstreams that are "hidden")
+  const derivedState = React.useMemo(() => deriveStateData(state), [state]);
```
some perf benefits to using `useMemo` for example, without it we'd be calling this func every time the user types in the text field

scarlettperry commented:
@danielhochman some more context on the recent commit: 

initially I passed down a list of projects that should be hidden as a prop to respective components (ie project group) so that they could filter it from the original state data. dropped the approach since we'd have to remember to filter out the final state data in each component

next i added a new `useState` hook for holding the the derived state value and initialized it with the original state data. in this way, the original state won't be replaced and we'd pass a copy of the final state (that doesn't contain the hidden upstreams/downstreams) to the components for final render. i also added a new `useEffect` hook to listen to changes in state in order to call the derivedstate func which sets the new value of the derived data. while that worked, seemed unnecessary to have a new useState and useEffect hook

in the current approach, I call the derivedstate func directly in the project selector which returns a copy of the state (without the hidden upstreams/downstreams) and I use that value in the StateContext which gets used for render in the project group component. i also tested this out with your data flow PR (value gets used in the new useEffect hook), and it worked out well.

danielhochman commented on frontend/workflows/projectSelector/src/helpers.tsx:
```diff
@@ -96,7 +97,92 @@ const exclusiveProjectDependencies = (
 };
 
+// evaluates whether an upstream/downstream should be hidden based on the checked status of/exclusivity to a project(s)
+const hidden = (
```
seems like it would be more efficient figure out the hidden projects (or conversely the ones that should be displayed) once for all selected projects rather than per project, lots of nested loops happening.

scarlettperry replied:
thx for raising! updated to remove unnecessary extra looping

scarlettperry commented on frontend/workflows/projectSelector/src/project-group.tsx:
```diff
@@ -140,7 +140,7 @@ const ProjectGroup: React.FC<ProjectGroupProps> = ({ title, group, displayToggle
               dispatch({
                 type: "TOGGLE_ENTIRE_GROUP",
-                payload: { group },
+                payload: { group, projects: Object.keys(state[group]) },
```
since we may sometimes render out filtered state data, we should only toggle what's shown in the group list. without it, we toggle the status of hidden objects since we currently toggle everything that is in the original state

danielhochman reviewed (approved):
i think it might be more efficient to compute the projects that should be displayed rather than computing the hidden projects. then you only need to loop over checked projects and push upstreams and downstreams to the list rather than looping over projects, then the dependency mappings as well, then removing upstream and downstreams.

however i don't think it's any less readable and the perf difference is probably negligible so ship it! great work!
